share/pvss.go

```diff
@@ -0,0 +1,174 @@
+package share
+
+import (
+	"errors"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/proof"
+	"github.com/dedis/crypto/random"
+)
+
+// This package implements public verifiable secret sharing as introduced by
+// Berry Schoenmakers at CRYPTO'99.
+
+// Some error definitions
+var errorTooFewShares = errors.New("not enough shares to recover secret")
+var errorDifferentLengths = errors.New("inputs of different lengths")
+
+// PubVerShare is a public verifiable share.
+type PubVerShare struct {
+	S PubShare        // Share
+	P proof.DLEQProof // Proof
+}
+
+// PVSS is the main public verifiable secret sharing struct.
+type PVSS struct {
+	suite abstract.Suite // Cryptographic suite
+	h     abstract.Point // Base point for polynomial commits
+	t     int            // Secret sharing threshold
+	n     int            // Number of shares
+}
+
+// NewPVSS creates a new PVSS struct using the given suite, base point, and
+// secret sharing threshold.
+func NewPVSS(s abstract.Suite, h abstract.Point, t int, n int) *PVSS {
+	return &PVSS{suite: s, h: h, t: t, n: n}
+}
+
+// EncShares creates encrypted PVSS shares using the public keys in X and
+// provides a NIZK encryption consistency proof for each share.
+func (pv *PVSS) EncShares(X []abstract.Point, secret abstract.Scalar) ([]*PubVerShare, *PubPoly, error) {
+
+	n := len(X)
+	encShares := make([]*PubVerShare, n)
+
+	// Create secret sharing polynomial
+	priPoly := NewPriPoly(pv.suite, pv.t, secret, random.Stream)
+
+	// Create secret set of shares
+	priShares := priPoly.Shares(n)
+
+	// Create public polynomial commitments with respect to basis H
+	pubPoly := priPoly.Commit(pv.h)
+
+	// Prepare data for encryption consistency proofs ...
+	indices := make([]int, n)
+	values := make([]abstract.Scalar, n)
+	H := make([]abstract.Point, n)
+	for i := 0; i < n; i++ {
+		indices[i] = priShares[i].I
+		values[i] = priShares[i].V
+		H[i] = pv.h
+	}
+
+	// Create NIZK discrete-logarithm equality proofs
+	proofs, _, sX, err := proof.NewDLEQProofBatch(pv.suite, H, X, values)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for i := 0; i < n; i++ {
+		ps := &PubShare{indices[i], sX[i]}
+		encShares[i] = &PubVerShare{*ps, *proofs[i]}
+	}
+
+	return encShares, pubPoly, nil
+}
+
+func (pv *PVSS) verifyEncShares(H abstract.Point, X []abstract.Point, polys []*PubPoly, encShares []*PubVerShare) ([]abstract.Point, []*PubVerShare, error) {
+
+	if len(X) != len(polys) && len(polys) != len(encShares) {
```

Comment:
question: do _both_ need to be different for an error, or is it enough if one pair is different? With this test, _both_ need to be different for the error.